### ISSUE TITLE
test_interpolator fails with quite a swing: -3.4 vs 3

### ISSUE THREAD

**Issue by yarikoptic** [nipy/nipy#468]
```
smaug:/tmp/buildd/nipy-0.4.3~rc1# python3 -m nose .pybuild/cpython3_3.9_nipy/build/nipy/algorithms/tests/test_interpolator.py:test_interpolator
nipy.algorithms.tests.test_interpolator.test_interpolator ... FAIL

======================================================================
FAIL: nipy.algorithms.tests.test_interpolator.test_interpolator
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/build/nipy-0.4.3~rc1/.pybuild/cpython3_3.9_nipy/build/nipy/algorithms/tests/test_interpolator.py", line 23, in test_interpolator
    assert_almost_equal(interp.evaluate([0, 0, 4]), arr[0, 0, -1])
  File "/usr/lib/python3/dist-packages/numpy/testing/_private/utils.py", line 579, in assert_almost_equal
    return assert_array_almost_equal(actual, desired, decimal, err_msg)
  File "/usr/lib/python3/dist-packages/numpy/testing/_private/utils.py", line 1042, in assert_array_almost_equal
    assert_array_compare(compare, x, y, err_msg=err_msg, verbose=verbose,
  File "/usr/lib/python3/dist-packages/numpy/testing/_private/utils.py", line 840, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 7 decimals

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 6.4
Max relative difference: 2.13333333
 x: array(-3.4)
 y: array(3)
>>  raise AssertionError('\nArrays are not almost equal to 7 decimals\n\nMismatched elements: 1 / 1 (100%)\nMax absolute difference: 6.4\nMax relative difference: 2.13333333\n x: array(-3.4)\n y: array(3)')
```

**Comment by yarikoptic** [comment]
relates to upgrade of scipy from 1.5.4-1+b1 to 1.6.0-2

**Comment by yarikoptic** [comment]
both this and #471 bisect to https://github.com/scipy/scipy/commit/81d5e7acf98a3215fd9280c2c1c0b8409756143f  ("Fix spline interpolation boundary handling for modes reflect and wrap") so now the question is either it is a regression or incorrect assumptions in nipy tests or functionality.

**Comment by effigies** [comment]
There was an issue related to that in nibabel that I believe has been fixed in the release, so I'm inclined to think we were maybe depending on incorrect behavior.

**Comment by yarikoptic** [comment]
I am yet to comprehend how interpolation is done there and how magically it worked before -- internally it does by default spline interpolation of order 3, so the data passed to scipy is already spline interpolated but then it does return correct values for the sample voxels I point to:
```python
diff --git a/nipy/algorithms/tests/test_interpolator.py b/nipy/algorithms/tests/test_interpolator.py
index 31870980..6ae92f39 100644
--- a/nipy/algorithms/tests/test_interpolator.py
+++ b/nipy/algorithms/tests/test_interpolator.py
@@ -19,7 +19,11 @@ def test_interpolator():
     coordmap =  vox2mni(np.eye(4))
     img = Image(arr, coordmap)
     # Interpolate off top right corner with different modes
-    interp = ImageInterpolator(img, mode='nearest')
+    interp = ImageInterpolator(img, order=3, mode='nearest')
+    data = interp.image.get_data()
+    for c in ([0, 0, 0], [0, 0, 1], [1,1,1]):
+        print(">>> ", c, data.item(*c), interp.evaluate(c))
+    print('>>>> ', interp.evaluate([0, 0, 4]))
     assert_almost_equal(interp.evaluate([0, 0, 4]), arr[0, 0, -1])
     interp = ImageInterpolator(img, mode='constant', cval=0)
     assert_array_equal(interp.evaluate([0, 0, 4]), 0)
```
which on old one returns
```
>>>  [0, 0, 1] 1 1.0000000000000002
>>>  [1, 1, 1] 17 17.000000000000007
>>>>  3.0000000000000018
ok
```
so output values correctly match original (up to minor flux) but then with new scipy it is completely off:
```
>>>  [0, 0, 1] 1 -5.999999999999999
>>>  [1, 1, 1] 17 23.0
>>>>  -3.399999999999993
FAIL
```

**Comment by matthew-brett** [comment]
The key difference is here:

```python
# Test 2D interpolation for Scipy
import numpy as np

import scipy.ndimage as spnd

arr = np.arange(6).reshape((2, 3))

coord = np.array([[0, 3]]).T
v1 = spnd.map_coordinates(arr,
                          coord,
                          order=3,
                          mode='nearest',
                         )
print('Calculate filter with interpolation', v1)

filtered = spnd.spline_filter(arr, order=3, mode='nearest')
v2 =  spnd.map_coordinates(filtered,
                           coord,
                           order=3,
                           mode='nearest',
                           prefilter=False)
print('Calculate filter followed by interpolation', v2)
```

On Scipy 1.2 this gives [2.] and [2.].   On Scipy 1.6.6 this gives [2.] and [2.20261874].  So it's the difference between using precalculated knots via `spline_filter`, or calculating the knots afresh within `map_coordinates`.

I'll file an issue with Scipy, and see what they say.

**Comment by matthew-brett** [comment]
I think this is fixed by #478 and #481 .
